src/roiextractors/extractors/tiffimagingextractors/multitiffmultipageextractor.py

```diff
@@ -0,0 +1,448 @@
+        self.dimension_order = dimension_order
+        self.channel_size = channel_size
+        self.depth_size = depth_size
+        self._sampling_frequency = sampling_frequency
+
+        # Get tifffile package
+        tifffile = get_package(package_name="tifffile")
+
+        # Open all TIFF files and store file handles for lazy loading
+        self._file_handles = []
+        total_pages = 0
+
+        for file_path in self.file_paths:
+            try:
+                tiff_handle = tifffile.TiffFile(file_path)
+                self._file_handles.append(tiff_handle)
+                total_pages += len(tiff_handle.pages)
+            except Exception as e:
+                # Close any opened file handles before raising the exception
+                for handle in self._file_handles:
+                    handle.close()
+                raise RuntimeError(f"Error opening TIFF file {file_path}: {e}")
+
+        # Get image dimensions from the first page of the first file
+        if not self._file_handles or not self._file_handles[0].pages:
+            raise ValueError("No valid TIFF files or pages found")
+
+        first_page = self._file_handles[0].pages[0]
+        self._num_rows, self._num_columns = first_page.shape
+        self._dtype = first_page.dtype
+
+        # Calculate time_size if not provided
+        frames_per_file = [len(handle.pages) for handle in self._file_handles]
+        total_frames = sum(frames_per_file)
+
+        if time_size is None:
+            # Calculate time_size based on total frames and other dimensions
+            frames_per_volume = channel_size * depth_size
+            if frames_per_volume == 0:
+                raise ValueError("Invalid dimension sizes: channel_size and depth_size cannot both be zero")
+
+            if total_frames % frames_per_volume != 0:
+                warnings.warn(
+                    f"Total frames ({total_frames}) is not divisible by frames per volume ({frames_per_volume}). "
+                    f"Some frames may not be accessible."
+                )
+
+            time_size = total_frames // frames_per_volume
+
+        self.time_size = time_size
+        self._num_frames = time_size
+
+        # Create mapping from frame index to file and page
+        self._frame_mapping = self._create_frame_mapping(
+            dimension_order=dimension_order,
+            channel_size=channel_size,
+            time_size=time_size,
+            depth_size=depth_size,
+            frames_per_file=frames_per_file,
+        )
+
+        # Store initialization parameters for potential reconstruction
+        self._kwargs = {
+            "file_paths": [str(path) for path in self.file_paths],
+            "dimension_order": dimension_order,
+            "channel_size": channel_size,
+            "time_size": time_size,
+            "depth_size": depth_size,
+            "sampling_frequency": sampling_frequency,
+        }
+
+    def _create_frame_mapping(
```

Comment:
This is the central part of the code where the mapping is done. Review this carefully. Check the same code on the ThorTiff extractor to see of a similar design pattern.